scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,312 @@
+  [new]
+  (let [cluster-definition-template ((config/compute-cluster-templates) (:template new))]
+    (assoc (cond
+             (not cluster-definition-template)
+             {:valid? false
+              :reason (str "Attempting to create cluster with unknown template: " (:template new))}
+             (not (:factory-fn cluster-definition-template))
+             {:valid? false
+              :reason (str "Template for cluster has no factory-fn: " cluster-definition-template)}
+             :else
+             {:valid? true})
+      :goal-config new :changed? true :cluster-name (:name new))))
+
+(defn check-for-unique-constraint-violations
+  "Check that the proposed resulting configurations don't collide on fields that should be unique, e.g. :base-path"
+  [changes resulting-active-configs unique-constraint-field]
+  (let [unique-constraint-value->cluster-names
+        (reduce (fn [m {:keys [cluster-name goal-config]}]
+                  (update m (goal-config unique-constraint-field) conj cluster-name))
+                {} resulting-active-configs)]
+    (->> changes
+         (map
+           #(let [{:keys [goal-config]} %
+                  unique-constraint-value (goal-config unique-constraint-field)
+                  clusters-for-unique-value (-> unique-constraint-value unique-constraint-value->cluster-names set)]
+              (cond-> %
+                (> (count clusters-for-unique-value) 1)
+                (assoc :valid? false
+                       :reason (str unique-constraint-field " is not unique between clusters " clusters-for-unique-value))))))))
+
+(defn compute-config-updates
+  "Take the current and desired configurations and compute the changes."
+  [db current-configs new-configs force?]
+  (let [[deletes-keys inserts-keys updates-keys] (util/diff-map-keys current-configs new-configs)
+        updates (->> (concat
+                       (map #(let [current (current-configs %)]
+                               (compute-config-update db current (assoc current :state :deleted) force?)) deletes-keys)
+                       (map #(compute-config-insert (new-configs %)) inserts-keys)
+                       (map #(compute-config-update db (current-configs %) (new-configs %) force?) updates-keys))
+                     (map (fn [{:keys [goal-config] :as update}] (assoc update :active? (not= :deleted (:state goal-config))))))
+        resulting-active-configs (->> updates (filter :valid?) (filter :active?))]
+    ;TODO check that number of running clusters is not less than a configured value. and make no changes if so
+    (-> updates
+      (check-for-unique-constraint-violations resulting-active-configs :base-path)
+      (check-for-unique-constraint-violations resulting-active-configs :ca-cert))))
+
+;TODO see if this is ok or need a better way
+(def scheduler-promise (promise))
+(def exit-code-syncer-state-promise (promise))
+
+(defn initialize-cluster!
+  "Create and initialize a ComputeCluster"
+  [config]
```

Comment:
Suggests using dictionary destructuring here:
 [{config-template :template config-config :config :as config}]  here

Side question: Shouldn't the key be :template-name, not :template? And :template-body instead of :template

Reply:
well, we picked template earlier. traded off some clarity for brevity. since that's such a top-level field

Reply:
updated